src/relay/pass/fuse_ops.cc

```diff
@@ -26,48 +697,74 @@ class SimpleFuser : public ExprMutator {
       return ExprMutator::VisitExpr_(fn_node);
     }
   }
-
+  // Transform calls.
   Expr VisitExpr_(const CallNode* call) {
     if (call->op.as<OpNode>()) {
-      // Placeholder fusion algorithm which abstracts
-      // single definitions into functions only.
-      Array<Var> params;
-      Array<Expr> inner_args;
-      Array<Expr> args;
-
-      int param_number = 0;
+      // If it is a primitive op call
+      // then we must have a group assignment for it already.
+      CHECK(gmap_.count(call));
+      auto* ret_group = gmap_.at(call)->FindRoot();
+      Array<Expr> new_args;
       for (auto arg : call->args) {
-        std::ostringstream os;
-        os << "p" << param_number++;
         auto type = arg->checked_type();
-        auto var = VarNode::make(os.str(), type);
-        params.push_back(var);
-        inner_args.push_back(var);
-        args.push_back(this->Mutate(arg));
+        CHECK(gmap_.count(arg.get()))
+            << "cannot find group of " << arg;
+        auto* arg_group = gmap_.at(arg.get())->FindRoot();
+        Expr new_arg = this->Mutate(arg);
+
+        if (ret_group != arg_group) {
+          Var param = ginfo_[ret_group].GetOrAllocParam(new_arg, type);
+          new_args.push_back(param);
+        } else {
+          new_args.push_back(new_arg);
+        }
+      }
+      auto new_call = CallNode::make(
+          call->op, new_args, call->attrs, call->type_args);
+
+      if (ret_group->root_ref == call) {
+        // This is the root of the group
+        // create the new call node.
+        const GroupInfo& ginfo = ginfo_[ret_group];
+        auto func = FunctionNode::make(
+            ginfo.params, new_call, call->checked_type(), {});
+        func = FunctionSetAttr(func, "Primitive", tvm::Integer(1));
+        return CallNode::make(func, ginfo.arguments, Attrs());
+      } else {
+        // This is an intermediate node of a fused function
+        // simply return the new call.
+        return new_call;
       }
-      auto body = CallNode::make(call->op, inner_args, call->attrs);
-      auto func = FunctionNode::make(
-          params, body, call->checked_type(), {});
-      func = FunctionSetAttr(func, "Primitive", tvm::Integer(1));
-      return CallNode::make(func, args, Attrs());
     } else {
       return ExprMutator::VisitExpr_(call);
     }
   }
+  // Debug function, dump the group assignment in text.
+  void DebugDumpGroup(const Expr& body) {
+    std::string text = RelayPrint(body, [this](const Expr& expr) -> std::string {
+        auto it = gmap_.find(expr.get());
+        if (it == gmap_.end()) return "";
+        std::ostringstream os;
+        auto *group = it->second->FindRoot();
+        os << "group=" << group;
+        return os.str();
+      });
+    LOG(INFO) << "Dump of group info:\n" << text;
+  }
 };
 
 
-Expr FuseOps(const Expr& expr) {
+Expr FuseOps(const Expr& expr, int fuse_opt_level) {
```

Comment:
I believe the pass should still take a module for the time being, and in the global var case recursively apply fusion. Our experiments with adding data types and using iterative structures requires that we apply these optimizations to all functions which are reachable from the main one.

Reply:
I agree,  however, I think we could wrap a FuseOps for envs via applies all, or get a list of functions via reachable analysis and then apply this Op. 

Since this pass optionally requires the env, I feel having a reachable analysis to pull out the list of functions then do foreach apply is a better approach